src/lib/core/decimal.h

```diff
@@ -37,52 +37,151 @@
+ * and named <struct decimal> by default.
+ *
+ * Define STRUCT_DECIMAL macro before inclusion of the module.h
+ * header to alias <decimal_t> to another structure. It should
+ * have the same data layout as one below: exactly same size (36)
+ * and alignment (4).
+ */
+#ifndef STRUCT_DECIMAL
+
+/**
+ * Decimal number.
+ *
+ * The data layout is the same as in the decNumber library.
+ */
+struct decimal {
+	/** Count of digits in the coefficient. Non-negative. */
+	int32_t digits;
+	/**
+	 * Unadjusted exponent, unbiased.
+	 *
+	 * It is in the range [-38; 37].
 	 */
-	DECIMAL_MAX_STR_LEN = DECIMAL_MAX_DIGITS + 14,
+	int32_t exponent;
+	/** Indicator bits. */
+	uint8_t bits;
+	/** Coefficient, from least significant unit. */
+	uint16_t lsu[13];
 };
 
-typedef decNumber decimal_t;
+#define STRUCT_DECIMAL struct decimal
+
+#endif
+
+typedef STRUCT_DECIMAL decimal_t;
+
+/** \endcond public */
+
+/*
+ * The documentation in the decnumber sources says the string needs to
+ * be >= digit count + 14.
+ */
+static_assert(DECIMAL_MAX_STR_LEN - DECIMAL_MAX_DIGITS == 14,
+	      "DECIMAL_MAX_STR_LEN - DECIMAL_MAX_DIGITS == 14");
+
+/*
+ * Data layout behind <decNumber> and <decimal_t> is transparent
+ * for a module API user. It means that we can't change it without
+ * ABI compatibility breakage.
+ *
+ * Say, if we'll add a field, the structures size in a compiled
+ * module and in tarantool will be different that may lead to out
+ * of buffer access. If the module will be recompiled against the
+ * new module.h header, it'll ABI incompatible with older
+ * tarantool versions.
+ *
+ * So, if the structure will going to change in tarantool,
+ * elaborate an upgrade pathway for module and application
+ * developers to minimize negative impact. Consider introduction
+ * of a new type and its own API instead of changing this one.
+ * Take care to equivalence of the data layout with cdata type
+ * offered by built-in decimal Lua module.
+ */
+static_assert(sizeof(decimal_t) == 36, "sizeof(decimat_t) == 36");
+static_assert(alignof(decimal_t) == 4, "alignof(decimat_t) == 4");
+
+/**
+ * Initialize a decimal with a value from the valid beginning
+ * of the string.
+ * If \a endptr is not NULL, store the address of the first
+ * invalid character in *endptr.
+ *
+ * If the number is less, than 10^DECIMAL_MAX_DIGITS,
+ * but has excess digits in fractional part, it will be rounded.
+ *
+ * @return NULL if string is invalid or
+ * the number is too big (>= 10^DECIMAL_MAX_DIGITS)
+ */
+decimal_t *
+strtodec(decimal_t *dec, const char *str, const char **endptr);
+
+/**
+ * Write the decimal to a string.
+ * Returns a statically allocated buffer containing
+ * the decimal representation.
+ */
+const char *
+decimal_str(const decimal_t *dec);
+
+/** \cond public */
 
 /**
  * @return decimal precision,
  * i.e. the amount of decimal digits in
  * its representation.
  */
-int
+API_EXPORT int
 decimal_precision(const decimal_t *dec);
```

Comment:
I don't think it's a good idea to expose our internal functions as the module API, because it limits our refactoring capabilities. For example, I don't like `decimal_t`, because we can't forward-declare it and have to include the huge libdecnumber header to declare functions that use it. I'd like to rework it as follows:

```C
struct decimal {
    decNumber impl;
};
```

But with this becoming a part of the public API I wouldn't be able to do that.

Also, shouldn't all module API functions have a `box_` or `tt_` prefix to avoid possible name collisions?